static/js/gamedata/legacy/ai.js

```diff
@@ -914,241 +937,228 @@ function AiSystemDestroy() {
-   }
-}
-
-function AiSystemExecute_FightOrMoveUnitsStepDone(player, unit) {
-   if (unit != undefined) {
-      UnitCounter++;
    }
-   AiSystemExecute_FightOrMoveUnitsStep(player);
-}
-
-function AiSystemExecute_StandardUnits(player, score, range) {
-   console.log("AiSystemExecute_StandardUnits()");
-   UnitCounter = 0;
-   AiSystemStandardUnitsDone = false;
-   AiSystemExecute_StandardUnitsStep(player, score, range);
+   UnitCounter++;
+   return didFight;
 }
 
 function AiSystemExecute_StandardUnitsStep(player, score, range) {
    if (UnitCounter >= compat.World.unitList.length) {
       AiSystemStandardUnitsDone = true;
-   }
-   else {
-      const unit = compat.World.unitList[UnitCounter];
-      if (unit.getPlayer().equals(player) && unit.getState() == Unit_State_WaitingForOrder && unit.getRange() > range) {
-         console.log("try to fight");
-         AiInfluencePrepareMap_Fight(unit);
-         AiInfluenceComputeMap(unit);
-         AiInfluenceFinalizeMap(unit);
-         const _loc2_ = gameMap_influence[unit.row][unit.col];
-         console.log("influence = " + _loc2_);
-         console.log("score = " + score);
-         if (_loc2_ >= score) {
-            AiToolsMove(unit, unit.getMovement());
-            // TODO: remove interval
-            AiSystemExecute_StandardUnitsINT = setInterval(AiSystemExecute_StandardUnitsWait, 500, unit, score, range);
-            return;
-         }
+      return false;
+   }
+   const unit = compat.World.unitList[UnitCounter];
+   if (unit.getPlayer().equals(player) && unit.getState() == Unit_State_WaitingForOrder && unit.getRange() > range) {
+      console.log("try to fight");
+      AiInfluencePrepareMap_Fight(unit);
+      AiInfluenceComputeMap(unit);
+      AiInfluenceFinalizeMap(unit);
+      const _loc2_ = gameMap_influence[unit.row][unit.col];
+      if (_loc2_ >= score) {
+         let didMove = AiToolsMove(unit, unit.getMovement());
+         StandardUnitMoving = false;
+         return didMove;
       }
-      UnitCounter++;
-      AiSystemExecute_StandardUnitsStep(player, score, range);
    }
+   UnitCounter++;
+   return false;
 }
 
-function AiSystemExecute_StandardUnitsWait(unit, score, range) {
-   if (!unit.isMoving) {
-      var _loc3_ = unit.player;
-      // TODO: remove interval
-      clearInterval(AiSystemExecute_StandardUnitsINT);
-      if (unit.getState() != Unit_State_Finished) {
-         console.log("attack");
-         AiToolsBattle(unit);
-         // TODO: remove interval
-         DisplayBattleINT = setTimeout(AiSystemExecute_StandardUnitsAttackDone, 4000, _loc3_, score, range, unit);
-         return;
+function AiSystemExecute_StandardUnitsWait(player, score, range) {
+   StandardUnitMoving = true;
+   const unit = compat.World.unitList[UnitCounter];
+   if (unit.getState() != Unit_State_Finished) {
+      console.log("attack");
+      let didFight = AiToolsBattle(unit);
+      if (unit.hp) {
+         UnitCounter++;

```

Comment:
What if the enemy unit died? Would we skip one?

Reply:
Or have index out of bounds in `const unit = compat.World.unitList[UnitCounter];`